src/main/java/com/atlassian/db/replica/api/AuroraMultiReplicaConsistency.java

```diff
@@ -61,12 +69,66 @@ public boolean isConsistent(Supplier<Connection> replicaSupplier) {
             });
     }
 
+    private boolean isConsistent_new(Supplier<Connection> replicaSupplier) {
+        return cluster.getReplicas(replicaSupplier)
+            .stream()
+            .allMatch(replica -> {
+                try (LazyConnectionSupplier connectionSupplier = new LazyConnectionSupplier(replica.getConnectionSupplier())) {
+                    return replicaConsistency.isConsistent(connectionSupplier);
+                } catch (ReadReplicaConnectionCreationException exception) {
+                    logger.warn("ReadReplicaConnectionCreationException occurred during consistency checking. It is likely that replica is the process of scaling, replica id: " + replica.getId(), exception);
+                    return true;
+                } catch (SQLException exception) {
+                    throw new ConnectionCouldNotBeClosedException(exception);
+                }
+            });
+    }
+
+    private static class LazyConnectionSupplier implements Supplier<Connection>, AutoCloseable {
+        final LazyReference<Connection> connectionLazyReference = new LazyReference<Connection>() {
+            @Override
+            protected Connection create() {
+                return supplier.get();
+            }
+        };
+        private final Supplier<Connection> supplier;
+
+        private LazyConnectionSupplier(Supplier<Connection> supplier) {
+            this.supplier = supplier;
+        }
+
+        /**
+         * The initial implementation of AuroraMultiReplicaConsistency was buggy. The implementation took the
+         * connection management responsibility from ReplicaConsistency. The current implementation fixes the issue
+         * by allowing ReplicaConsistency implementation to take full responsibility for the connection lifecycle.
+         *
+         * Unfortunately, the existing implementations of ReplicaConsistency may rely on the bug (for example some
+         * implementations in our product rely on it). To make it backwards compatible (behavioural compatibility),
```

Comment:
I don't get it, can you share an example?
If this is product-oriented you can share it via our slack channel if you prefer.

Reply:
`SequenceReplicaConsistency#isConsistentPerReplicaLsn` gets the consistency from the supplier, but never closes it. It relies on the bug to release the connection to the connection pool.

Reply:
I see, do you plan to fix it under this issue, or we should create a new one?

Maybe in product `isConsistentPerReplicaLsn` we should add `ifNotClosed -> close`

Reply:
I'm going to fix the issue in our product, but I don't know how other products use it, so I keep it compatible just in case.

Reply:
👍